Page List: Show empty placeholder if no items

mkaz opened the pull request:
## Description

If no pages exist the Page List block will be empty, making it impossible to select in the rendered view in editor. This PR adds an empty placeholder to show where the block would be and allows for selecting in rendered view.

The Page List block can be empty if a user deletes all pages, or while working on #31471 there may be no children when showing a child list.

## How has this been tested?

- Add a Page List to a post or page
- Confirm it shows fine
- Delete all pages from your site
- View post with Page List block and confirm it is not shown
- Apply PR change
- View post again and confirm page list shows

## Screenshots <!-- if applicable -->

![image](https://user-images.githubusercontent.com/45363/136471423-2ed10b57-3db5-47cc-99ee-a27a0c193e20.png)

## Types of changes

- Add empty response placeholder for empty server-side render response
- Updates render_callback to return early and empty if no top level pages

mkaz committed:
Add empty response placeholder

mkaz committed:
Return early if no top level pages

mkaz commented:
@jasmussen I use just the default `<Placeholder/>` component but it doesn't render with nice padding. I can add some in, or other suggestion for design on the placeholder?

tellthemachines commented on packages/block-library/src/page-list/index.php:
```diff
@@ -281,6 +281,12 @@ function render_block_core_page_list( $attributes, $content, $block ) {
+	// Return early and empty to trigger EmptyResponsePlaceholder.
+	if ( empty( $top_level_pages ) ) {
+		return;
+	}
```
We could return even earlier if we instead checked `$all_pages` above.

tellthemachines reviewed (approved):
Looks good! Just a minor comment below, but not a blocker.

jasmussen commented:
Nice one! Before:
<img width="781" alt="Screenshot 2021-10-08 at 10 16 33" src="https://user-images.githubusercontent.com/1204802/136522671-c6207a7a-3179-4b34-9e94-07d474cb268b.png">

After:

<img width="1031" alt="Screenshot 2021-10-08 at 10 21 52" src="https://user-images.githubusercontent.com/1204802/136523120-3468dbf2-05fd-4497-98f4-cbe7429c6694.png">

_The first is the page list inside navigation, the second is just a standalone page list._

I also confirmed that the placeholder is only shown in the editor.

_However_, in light of what we're doing to other FSE blocks (see #35243 and #35242), I think we should avoid the use of the `Placeholder` component entirely. It feels entirely sufficient to output just the text phrase, and inherit everything else from the theme. That way it's substantially easier for a theme developer to develop their theme, to see that styles are correctly inherited, and to not have to worry about the clunky placeholder and the editor styles it brings with it isn't the source of trouble.

Removing the placeholder could make it look like this:
<img width="694" alt="Screenshot 2021-10-08 at 10 24 42" src="https://user-images.githubusercontent.com/1204802/136523507-19a66c3a-bc1f-4375-b448-5ae43f11ca4f.png">

This is the markup I inspected myself to:

<img width="664" alt="Screenshot 2021-10-08 at 10 24 49" src="https://user-images.githubusercontent.com/1204802/136523529-461bd276-afdd-46b7-8447-ce75024da98d.png">

mkaz committed:
Simplify placeholder to just span and text

mkaz committed:
Move the check for empty pages up to all_pages

This makes it more clear and bails out even earlier.

mkaz commented:
Thanks @tellthemachines good catch, I moved the check up even earlier, which also makes it more clear.

Thanks @jasmussen, I simplified the placeholder to just that span tag. 👍

jasmussen commented:
Nice! Ship it.

jasmussen commented:
@mkaz Two side effects of this one:

1. The spinner now seems to appear very often (see GIF below)
2. When first opening up the Site Editor, the Page List will show "Page List: No pages." for a very long time while loading.

GIF showing the message, and the spinner recurring:
![spinner](https://user-images.githubusercontent.com/1204802/137299072-e3c78f49-84c8-4843-bf70-3990264568d0.gif)

This makes the block feel a bit brittle. Can we follow up with a few tweaks?

- Never show the spinner in this block, it shouldn't be doing any loading work after the initial load.
- Delay the "No pages" message until we're absolutely sure there really are no pages.
